robot_3d/src/joint_states_subscriber_3d.cpp

```diff
@@ -0,0 +1,233 @@
+#include <ros/ros.h>
+#include "ros/package.h"
+#include <sensor_msgs/JointState.h>
+#include "std_msgs/String.h"
+
+#include <tf/transform_broadcaster.h>
+#include <tf/transform_datatypes.h>
+
+#include <kdl_parser/kdl_parser.hpp>
+#include <kdl/chain.hpp>
+#include <kdl/chainfksolver.hpp>
+#include <kdl/chainfksolverpos_recursive.hpp>
+#include <kdl/frames_io.hpp>
+
+#include <tf_conversions/tf_kdl.h>
+
+#include <boost/filesystem.hpp>
+
+#include "yaml-cpp/yaml.h"
+
+#include <string>
+#include <sstream>
+#include <vector>
+
+class JointMeasurementData
+{
+public:
+  JointMeasurementData(ros::NodeHandle* n);
+
+private:
+  void jointStatesCallback(const sensor_msgs::JointState::ConstPtr& msg);
+  // Method to calculate the end effector position using kdl and broadcast the transform.
+  void kdlCalc(double joint_positions[]);
+  // Method to output the pose of the end effector position to yaml file.
+  void outputYaml(tf::Transform kdl_transform, double joint_positions[]);
+  std::vector<double> getJointOffsets(double joint_positions[]);
+
+  ros::Subscriber sub_;
+
+  // String to store the filepath for yaml file.
+  std::string output_filepath_;
+  // String to store data to be sent to yaml file.
+  std::string output_data_ = "Data: \n";
+  // The number of data points accepted by node until it terminates.
+  int num_data_point_;
+  // Count of the data points.
+  int data_point_count_;
+};
+
+JointMeasurementData::JointMeasurementData(ros::NodeHandle* n)
+{
+  // Sets the count to zero.
+  data_point_count_ = 0;
+
+  sub_ = n->subscribe("joint_states", 1000, &JointMeasurementData::jointStatesCallback, this);
+
+  // Gets the maximum number of data points from launch file.
+  n->getParam("data_point_count", num_data_point_);
+
+  // Gets the currrent date and time to include in the output filename.
+  char date_holder[21];
+  time_t now;
+  now = time(0);
+  if (now != -1)
+  {
+    strftime(date_holder, 20, "%d_%m_%Y_%T", gmtime(&now));
+  }
+
+  // Gets the robot_3d package filepath.
+  output_filepath_ = ros::package::getPath("robot_3d");
+
+  // Checks if Output_yaml folder exists in package and creates folder if it does not already exist.
+  std::string check = output_filepath_.append("/Output_yaml");
```

Comment:
the .append method operates and changes the output_filepath_ IN PLACE. So there is no need for the additional `check `variable

Reply:
Does that mean that every time the program is run a new output_yaml folder will be created?

Reply:
Nah. I'm saying that after you use `.append()`, you don't need to reassign the variable `output_filepath_` to `check` because the `.append()` changes `output_filepath_` immediately.